pkg/middleware/region_proxy.go

```diff
@@ -46,58 +77,134 @@ func NewRegionProxy(ctx context.Context, options ...RegionProxyMiddwareOption) *
+	}
+
+	if regionProxyMiddleware.logger == nil {
+		regionProxyMiddleware.logger, _ = sdk.NewGoLoggerBuilder().
 			Info(true).
 			Build()
 	}
 
-	if regionProxyMiddleware.GetDispatchHostFunc == nil {
-		regionProxyMiddleware.Logger.Warn(ctx, "GetDispatchHostFunc is not defined for the region proxy")
+	if regionProxyMiddleware.dispatchHandler == nil {
+		regionProxyMiddleware.dispatchHandler = defaultDispatchHandler()
 	}
 
-	if regionProxyMiddleware.ErrorHandler == nil {
-		regionProxyMiddleware.ErrorHandler = func(w http.ResponseWriter, r *http.Request, err error) {
-			w.WriteHeader(http.StatusInternalServerError)
-			w.Write([]byte(fmt.Sprintf("Error in region proxy: %v", err)))
-		}
+	if regionProxyMiddleware.errorHandler == nil {
+		regionProxyMiddleware.errorHandler = defaultErrorHandler()
 	}
 
 	return regionProxyMiddleware
 }
 
 func (rp *RegionProxy) Handler(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if rp.GetDispatchHostFunc == nil {
-			next.ServeHTTP(w, r)
-			return
-		}
+		var err error
 		ctx := r.Context()
-		dispatchHost := ""
-		dispatchHost, err := rp.GetDispatchHostFunc(r.Context(), rp.Logger, r, rp.Connection)
-		if err != nil {
-			rp.ErrorHandler(w, r, err)
+		ids := rp.getClusterIdsHandler(ctx, rp.logger, r)
+		if ids.Id == "" && ids.ExternalId == "" {
+			rp.logger.Warn(ctx, "Cannot find cluster id or external id from request")
+			next.ServeHTTP(w, r)
 			return
 		}
 
-		if dispatchHost != "" {
-			rp.Logger.Info(ctx, "Dispatch the request %s to %s", r.URL, dispatchHost)
-			requestsDispatched.Inc()
-			dispatchURL, err := url.Parse(dispatchHost)
+		rhRegionId, found := rp.checkCache(ids.Id, ids.ExternalId)
+		if !found {
+			if rp.checkLocalHandler != nil {
+				exists, err := rp.checkLocalHandler(ctx, rp.logger, ids)
+				if err != nil {
+					rp.errorHandler(w, r, err)
+					return
+				}
+				if exists {
+					rp.updateCache("", ids.Id, ids.ExternalId)
+					next.ServeHTTP(w, r)
+					return
+				}
+			}
+			rhRegionId, err = getRhRegionId(ctx, rp.logger, ids, rp.connection)
+			rp.updateCache(rhRegionId, ids.Id, ids.ExternalId)
 			if err != nil {
-				rp.ErrorHandler(w, r, err)
+				rp.errorHandler(w, r, err)
 				return
 			}
-			if dispatchURL.Scheme == "" {
-				dispatchURL = &url.URL{
-					Host:   dispatchHost,
-					Scheme: "https"}
-			}
-			r.Host = dispatchURL.Host
-			proxy := httputil.NewSingleHostReverseProxy(dispatchURL)
-			proxy.ServeHTTP(w, r)
-		} else {
-			next.ServeHTTP(w, r)
+		}
+
+		err = rp.dispatchHandler(ctx, rp.logger, w, r, next, rhRegionId, rp.regions)
+		if err != nil {
+			rp.errorHandler(w, r, err)
 		}
 	})
 }
+
+func defaultDispatchHandler() dispatchHandlerFunc {
+	return func(ctx context.Context, logger logging.Logger, w http.ResponseWriter, r *http.Request,
+		next http.Handler, rhRegionId string, regions map[string]sdk.Region) error {
+		if !strings.Contains(rhRegionId, APSoutEast1) {
```

Comment:
Wondering if this could be made more generic instead of hardcoding the region check

Reply:
yes, this is the current default behavior for the dispatch behavior, which only handles the ap-southeast-1 region.
Once we have more required proxy scenarios, we can re-visit this part

Reply:
I don't think ocm-common ever cares about the key represented by `APSouthEast1`.  There is no reason to hardcode this check.